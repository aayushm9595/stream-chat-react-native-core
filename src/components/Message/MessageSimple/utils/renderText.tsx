import React from 'react';
// @ts-expect-error
import Markdown from '@stream-io/react-native-simple-markdown';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';

import type { MarkdownStyle } from '../../../../styles/themeConstants';
import type { Message } from '../../../MessageList/utils/insertDates';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../../types/types';

const defaultMarkdownStyles = {
  inlineCode: {
    backgroundColor: '#F3F3F3',
    borderColor: '#dddddd',
    color: 'red',
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  link: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
  url: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
};

export type RenderTextParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  markdownRules: UnknownType;
  markdownStyles: MarkdownStyle;
  message: Message<At, Ch, Co, Ev, Me, Re, Us>;
};

export const renderText = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: RenderTextParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { markdownRules, markdownStyles, message } = params;

  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users = [], text } = message;

  if (!text) return null;

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 200,
      omission: '...',
    });
    const markdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, markdown);
  }

  if (mentioned_users.length) {
    for (let i = 0; i < mentioned_users.length; i++) {
      const username = mentioned_users[i].name || mentioned_users[i].id;
      const markdown = `**@${username}**`;
      const regEx = new RegExp(`@${username}`, 'g');
      newText = newText.replace(regEx, markdown);
    }
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
  };

  return (
    <Markdown rules={markdownRules} styles={styles}>
      {newText}
    </Markdown>
  );
};
