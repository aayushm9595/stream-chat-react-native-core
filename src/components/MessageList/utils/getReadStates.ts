import {
  InsertDatesResponse,
  isDateSeparator,
  Message,
} from '../utils/insertDates';

import type { ImmutableDate } from 'seamless-immutable';
import type { UserResponse } from 'stream-chat';

import type { ChannelContextValue } from '../../../contexts/channelContext/ChannelContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const getReadStates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us>,
  read?: ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>['read'],
) => {
  const readData = messages.reduce((acc, cur) => {
    if (!isDateSeparator(cur) && cur.id) {
      acc[cur.id] = [];
    }
    return acc;
  }, {} as { [key: string]: UserResponse<Us>[] });

  const filteredMessagesReversed = messages
    .filter((msg) => !isDateSeparator(msg) && msg.updated_at)
    .reverse() as Array<
    Message<At, Ch, Co, Ev, Me, Re, Us> & {
      updated_at: string | ImmutableDate;
    }
  >;

  if (read) {
    /**
     * Channel read state is stored by user
     */
    for (const readState of Object.values(read)) {
      /**
       * If no last read continue
       */
      if (!readState.last_read) {
        continue;
      }

      /**
       * Array is in reverse order so newest message is at 0,
       * we find the index of the first message that is older
       * than the last read and then set last read to that, or
       * if there are no newer messages, the first message is
       * last read message.
       */
      const userLastReadMsgId = filteredMessagesReversed.find(
        (msg) => msg.updated_at < readState.last_read,
      )?.id;

      /**
       * If there there is a last read message add the user
       * to the array of last reads for that message
       */
      if (userLastReadMsgId) {
        readData[userLastReadMsgId] = [
          ...readData[userLastReadMsgId],
          readState.user as UserResponse<Us>,
        ];
      }
    }
  }

  return readData;
};
