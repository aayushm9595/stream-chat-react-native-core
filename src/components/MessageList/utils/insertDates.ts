import type { ChannelState } from 'stream-chat';

import type {
  MessagesContextValue,
  MessageWithDates,
} from '../../../contexts/messagesContext/MessagesContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type DateSeparator<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  date: Message<At, Ch, Co, Ev, Me, Re, Us>['created_at'] | string;
  type: 'message.date';
};

export type MessageOrDate<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | Message<At, Ch, Co, Ev, Me, Re, Us>
  | DateSeparator<At, Ch, Co, Ev, Me, Re, Us>;

export type InsertDatesResponse<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>[];

export const isDateSeparator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  message: MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>,
): message is DateSeparator<At, Ch, Co, Ev, Me, Re, Us> =>
  message.type === 'message.date';

export type Message<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | ReturnType<ChannelState<At, Ch, Co, Ev, Me, Re, Us>['messageToImmutable']>
  | MessageWithDates<At, Ch, Co, Me, Re, Us>;

export const insertDates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages:
    | MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'],
) => {
  const newMessages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us> = [];
  if (messages.length === 0) {
    return newMessages;
  }

  for (const [i, message] of messages.entries()) {
    /**
     * If message read or deleted don't consider for date labels
     */
    if (message.type === 'message.read' || message.deleted_at) {
      newMessages.push(message);
      continue;
    }

    /**
     * Get the date of the current message and create
     * variable for previous date (day)
     */
    const messageDate = message.created_at.getDay();

    /**
     * If this is not the last entry in the messages array
     * set the previous message date (day) to the date of the next
     * message in the array
     */
    const prevMessageDate =
      i < messages.length - 1
        ? messages[i + 1].created_at.getDay()
        : messageDate;

    /**
     * Before the first message insert a date object
     */
    if (i === 0) {
      newMessages.push(
        {
          date: message.created_at,
          type: 'message.date',
        } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>,
        message,
      );

      /**
       * If the date (day) has changed between two messages
       * insert a date object
       */
    } else if (messageDate !== prevMessageDate) {
      newMessages.push(message, {
        date: messages[i + 1].created_at,
        type: 'message.date',
      } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>);

      /**
       * Otherwise just add the message
       */
    } else {
      newMessages.push(message);
    }
  }

  return newMessages;
};
