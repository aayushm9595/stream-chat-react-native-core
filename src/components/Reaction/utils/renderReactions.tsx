import React from 'react';
import { Text } from 'react-native';

import type { ReactionResponse } from 'stream-chat';

import type { LatestReactions, Reaction } from '../ReactionList';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const renderReactions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  reactions: LatestReactions<At, Ch, Co, Me, Re, Us>,
  supportedReactions: Reaction[],
) => {
  const reactionsByType: {
    [key: string]: ReactionResponse<Re, Us>[];
  } = {};
  reactions?.map((reaction) => {
    if (!reactionsByType[reaction.type]) {
      return (reactionsByType[reaction.type] = [reaction]);
    }
    return (reactionsByType[reaction.type] = [
      ...reactionsByType[reaction.type],
      reaction,
    ]);
  });

  const emojiDataByType: { [key: string]: Reaction } = {};
  const reactionTypes = supportedReactions.map((supportedReaction) => {
    emojiDataByType[supportedReaction.id] = supportedReaction;
    return supportedReaction.id;
  });

  return Object.keys(reactionsByType).map((type) =>
    reactionTypes.indexOf(type) > -1 ? (
      <Text key={type} testID={type}>
        {emojiDataByType[type].icon}
      </Text>
    ) : null,
  );
};
