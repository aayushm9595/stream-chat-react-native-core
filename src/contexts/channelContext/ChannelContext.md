The channel context exposes the following properties:

- **channel** {object} The currently active chat channel
- **disabled** {boolean} Whether or not the channel UI is frozen/disabled
- **EmptyStateIndicator** {component} UI component for empty Channel state
- **error** {boolean} Whether or not the channel errored on loading
- **eventHistory** {object} History of all native events received during a session
- **lastRead** {Date} Latest initialization of the channel
- **loading** {boolean} Whether or not the channel is currently loading
- **markRead** {function} Marks the current channel as read, runs on channel mount
- **members** {ImmutableObject} All members of the current channel

**Example:**

```json
{
  "thierry-123": {
    "id": "thierry-123",
    "role": "user",
    "created_at": "2019-04-03T14:42:47.087869Z",
    "updated_at": "2019-04-16T09:20:03.982283Z",
    "last_active": "2019-04-16T11:23:51.168113408+02:00",
    "online": true
  },
  "vishal-123": {
    "id": "vishal-123",
    "role": "user",
    "created_at": "2019-05-03T14:42:47.087869Z",
    "updated_at": "2019-05-16T09:20:03.982283Z",
    "last_active": "2019-06-16T11:23:51.168113408+02:00",
    "online": false
  }
}
```
- **read** {ImmutableObject} The read state for each user
- **typing** {ImmutableObject} Map of user IDs for currently typing users, corresponds to the `typing.start` [event](https://getstream.io/chat/docs/#event_object)

**Example:**

```json
{
  "user_id_1": typing_event_object_of_user_1,
  "user_id_2": typing_event_object_of_user_2
}
```
- **watcher_count** {number} Number of members watching the channel
- **watchers** {ImmutableObject} Map of user IDs for users currently online and watching the channel

**Example:**

```json
{
  "thierry-123": {
    "id": "thierry-123",
    "role": "user",
    "created_at": "2019-04-03T14:42:47.087869Z",
    "updated_at": "2019-04-16T09:20:03.982283Z",
    "last_active": "2019-04-16T11:23:51.168113408+02:00",
    "online": true
  },
  "vishal-123": {
    "id": "vishal-123",
    "role": "user",
    "created_at": "2019-05-03T14:42:47.087869Z",
    "updated_at": "2019-05-16T09:20:03.982283Z",
    "last_active": "2019-06-16T11:23:51.168113408+02:00",
    "online": true
  }
}
```

