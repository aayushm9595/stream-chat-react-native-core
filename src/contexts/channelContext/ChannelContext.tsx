import React, { PropsWithChildren, useContext } from 'react';

import { getDisplayName } from '../utils/getDisplayName';

import type { Channel, ChannelState, Event } from 'stream-chat';

import type { EmptyStateProps } from '../../components/Indicators/EmptyStateIndicator';
import type { LoadingProps } from '../../components/Indicators/LoadingIndicator';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

export type ChannelContextValue<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  EmptyStateIndicator: React.ComponentType<EmptyStateProps>;
  error: boolean;
  eventHistory: { [key: string]: Event<At, Ch, Co, Ev, Me, Re, Us>[] };
  loading: boolean;
  LoadingIndicator: React.ComponentType<LoadingProps>;
  markRead: () => void;
  members: ChannelState<At, Ch, Co, Ev, Me, Re, Us>['members'];
  read: ChannelState<At, Ch, Co, Ev, Me, Re, Us>['read'];
  setLastRead: React.Dispatch<React.SetStateAction<Date | undefined>>;
  typing: ChannelState<At, Ch, Co, Ev, Me, Re, Us>['typing'];
  watchers: ChannelState<At, Ch, Co, Ev, Me, Re, Us>['watchers'];
  channel?: Channel<At, Ch, Co, Ev, Me, Re, Us>;
  disabled?: boolean;
  lastRead?: Date;
  watcherCount?: ChannelState<At, Ch, Co, Ev, Me, Re, Us>['watcher_count'];
};

export const ChannelContext = React.createContext({} as ChannelContextValue);

export const ChannelProvider = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>({
  children,
  value,
}: PropsWithChildren<{
  value: ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>;
}>) => (
  <ChannelContext.Provider value={(value as unknown) as ChannelContextValue}>
    {children}
  </ChannelContext.Provider>
);

export const useChannelContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>() =>
  (useContext(ChannelContext) as unknown) as ChannelContextValue<
    At,
    Ch,
    Co,
    Ev,
    Me,
    Re,
    Us
  >;

/**
 * Typescript currently does not support partial inference so if ChatContext
 * typing is desired while using the HOC withChannelContext the Props for the
 * wrapped component must be provided as the first generic.
 */
export const withChannelContext = <
  P extends UnknownType,
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  Component: React.ComponentType<P>,
): React.FC<Omit<P, keyof ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>>> => {
  const WithChannelContextComponent = (
    props: Omit<P, keyof ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>>,
  ) => {
    const channelContext = useChannelContext<At, Ch, Co, Ev, Me, Re, Us>();

    return <Component {...(props as P)} {...channelContext} />;
  };
  WithChannelContextComponent.displayName = `WithChannelContext${getDisplayName(
    Component,
  )}`;
  return WithChannelContextComponent;
};
