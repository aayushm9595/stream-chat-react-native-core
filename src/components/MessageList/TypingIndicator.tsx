import React from 'react';

import { useTypingString } from './hooks/useTypingString';

import { AvatarProps, Avatar as DefaultAvatar } from '../Avatar/Avatar';

import { useChannelContext } from '../../contexts/channelContext/ChannelContext';
import { useChatContext } from '../../contexts/chatContext/ChatContext';
import { styled } from '../../styles/styledComponents';

import type { Event, UserResponse } from 'stream-chat';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

const Container = styled.View`
  align-items: flex-end;
  flex-direction: row;
  justify-content: flex-start;
  ${({ theme }) => theme.typingIndicator.container.css};
`;

const TypingText = styled.Text`
  color: ${({ theme }) => theme.typingIndicator.text.color};
  font-size: ${({ theme }) => theme.typingIndicator.text.fontSize}px;
  margin-left: 10px;
  ${({ theme }) => theme.typingIndicator.text.css};
`;

export type TypingIndicatorProps = {
  /**
   * Defaults to and accepts same props as: [Avatar](https://getstream.github.io/stream-chat-react-native/#avatar)
   */
  Avatar?: React.ComponentType<AvatarProps>;
};

export const TypingIndicator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType
>({
  Avatar = DefaultAvatar,
}: TypingIndicatorProps) => {
  const { typing } = useChannelContext<At, Ch, Co, Ev, Me, Re, Us>();
  const { client } = useChatContext<At, Ch, Co, Ev, Me, Re, Us>();
  const typingString = useTypingString<At, Ch, Co, Ev, Me, Re, Us>();

  const typingUsers = Object.values(typing);

  return (
    <Container testID='typing-indicator'>
      {(typingUsers.filter(
        ({ user }) => !!user && user.id !== client?.user?.id,
      ) as Array<
        Event<At, Ch, Co, Ev, Me, Re, Us> & { user: UserResponse<Us> }
      >).map(({ user }, idx) => (
        <Avatar
          image={user.image}
          key={`${user.id}${idx}`}
          name={user.name || user.id}
          size={20}
          testID={`typing-avatar-${idx}`}
        />
      ))}
      <TypingText>{typingString}</TypingText>
    </Container>
  );
};
