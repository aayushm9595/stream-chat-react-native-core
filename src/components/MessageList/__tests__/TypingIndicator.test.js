import React from 'react';
import { View } from 'react-native';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import { generateStaticUser, generateUser } from 'mock-builders/generator/user';
import { getTestClientWithUser } from 'mock-builders/mock';

import { TypingIndicator } from '../TypingIndicator';

import { Chat } from '../../Chat/Chat';

import { ChannelProvider } from '../../../contexts/channelContext/ChannelContext';
import { TranslationProvider } from '../../../contexts/translationContext/TranslationContext';
import { Streami18n } from '../../../utils/Streami18n';

afterEach(cleanup);

describe('TypingIndicator', () => {
  let chatClient;

  it('should render typing indicator for two users', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();
    const user2 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 }, user2: { user: user2 } };

    const { getByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );
    expect(t).toHaveBeenCalledWith(
      '{{ firstUser }} and {{ secondUser }} are typing...',
      {
        firstUser: user1.name,
        secondUser: user2.name,
      },
    );
    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
    });
  });

  it('should render typing indicator for one user', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 } };

    const { getByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );
    expect(t).toHaveBeenCalledWith('{{ user }} is typing...', {
      user: user1.name,
    });
    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
    });
  });

  it('should render typing Avatar for one user', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 } };

    const { getByTestId, queryByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator Avatar={View} />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );
    expect(t).toHaveBeenCalledWith('{{ user }} is typing...', {
      user: user1.name,
    });

    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
      expect(getByTestId('typing-avatar-0')).toBeTruthy();
      expect(queryByTestId('typing-avatar-1')).toBeFalsy();
    });
  });

  it('should render typing Avatar for two users', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();
    const user2 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 }, user2: { user: user2 } };

    const { getByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator Avatar={View} />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
      expect(getByTestId('typing-avatar-0')).toBeTruthy();
      expect(getByTestId('typing-avatar-1')).toBeTruthy();
    });
  });

  it('should render no typing Avatars', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = {};

    const { getByTestId, queryByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator Avatar={View} />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
      expect(queryByTestId('typing-avatar-0')).toBeFalsy();
      expect(queryByTestId('typing-avatar-1')).toBeFalsy();
    });
  });

  it('should match typing indicator snapshot', async () => {
    const i18nInstance = new Streami18n();
    const { t } = await i18nInstance.getTranslators();
    const user0 = generateStaticUser(0);
    const user1 = generateStaticUser(1);
    const user2 = generateStaticUser(3);

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 }, user2: { user: user2 } };

    const { toJSON } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <ChannelProvider value={{ typing }}>
            <TypingIndicator />
          </ChannelProvider>
        </TranslationProvider>
      </Chat>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
