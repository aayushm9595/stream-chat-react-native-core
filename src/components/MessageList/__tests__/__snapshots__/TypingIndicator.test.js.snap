// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypingIndicator should match typing indicator snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "flex-start",
      },
    ]
  }
  testID="typing-indicator"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Image
      accessibilityLabel="initials"
      onError={[Function]}
      resizeMethod="resize"
      size={20}
      source={
        Object {
          "uri": "https://i.imgur.com/SLx06PP.png",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 20,
            "width": 20,
          },
        ]
      }
      testID="typing-avatar-0"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Image
      accessibilityLabel="initials"
      onError={[Function]}
      resizeMethod="resize"
      size={20}
      source={
        Object {
          "uri": "https://i.imgur.com/T68W8nR_d.webp?maxwidth=728&fidelity=grand",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 20,
            "width": 20,
          },
        ]
      }
      testID="typing-avatar-1"
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0,0,0,0.5)",
          "fontSize": 14,
          "marginLeft": 10,
        },
      ]
    }
  >
    Grant and Neil are typing...
  </Text>
</View>
`;
